pkg/receive/writer.go

```diff
@@ -81,6 +82,14 @@ func (r *Writer) Write(ctx context.Context, tenantID string, wreq *prompb.WriteR
 			lset = labelpb.ZLabelsToPromLabels(t.Labels)
 		}
 
+		// If the ref is 0, it indicates that inserting the samples will create a new time series.
+		// We do the seriesLimit check before actually creating the series in head because even after
+		// a rollback, the series will stay in head, defeating the whole purpose of series limit.
+		if ref == 0 && seriesLimit > 0 && s.NumSeries() >= seriesLimit {
```

Comment:
Can we move this to the for loop on line 78? I think we don't need to allocate memory for labels in this case.